src/main/scala/mesosphere/marathon/core/health/impl/AppHealthCheckActor.scala

```diff
@@ -0,0 +1,157 @@
+  case class InstanceKey(applicationKey: ApplicationKey, instanceId: Instance.Id)
+
+  def props(eventBus: EventStream): Props = Props(new AppHealthCheckActor(eventBus))
+
+  case class AddHealthCheck(appKey: ApplicationKey, healthCheck: HealthCheck)
+  case class RemoveHealthCheck(appKey: ApplicationKey, healthCheck: HealthCheck)
+
+  case class PurgeHealthCheckStatuses(hc: Seq[(InstanceKey, HealthCheck)])
+  case class HealthCheckStatusChanged(
+      appKey: ApplicationKey,
+      healthCheck: HealthCheck, health: Health)
+
+  trait InstanceHealthChangedNotifier {
+    def notify(applicationKey: ApplicationKey, instanceId: Instance.Id,
+      healthiness: Option[Boolean]): Unit
+  }
+
+  class AppHealthCheckProxy extends StrictLogging {
+    /**
+      * Map of health check definitions of all applications
+      */
+    var healthChecks: mutable.Map[ApplicationKey, Set[HealthCheck]] = mutable.Map.empty
+
+    /**
+      *  Map of results of all health checks for all applications.
+      *  Results are optional, therefore the global health status of an instance
+      *  is either:
+      *    unknown (if some results are still missing)
+      *    healthy (if all results are known and healthy)
+      *    not healthy (if all results are known and at least one is unhealthy)
+      */
+    var healthCheckStates: mutable.Map[InstanceKey, Map[HealthCheck, Option[Health]]] =
+      mutable.Map.empty
+
+    private def computeGlobalHealth(instanceHealthResults: Map[HealthCheck, Option[Health]]): Option[Boolean] = {
+      val isHealthAlive = (health: Option[Health]) => health.fold(false)(_.alive)
+      val isHealthUnknown = (health: Option[Health]) => health.isEmpty
+
+      if (instanceHealthResults.values.forall(isHealthAlive))
+        Some(true)
+      else if (instanceHealthResults.values.exists(isHealthUnknown))
+        Option.empty[Boolean]
+      else
+        Some(false)
+    }
+
+    def addHealthCheck(applicationKey: ApplicationKey, healthCheck: HealthCheck): Unit = {
+      healthChecks.update(applicationKey, healthChecks.getOrElse(applicationKey, Set.empty) + healthCheck)
+      logger.debug(s"Add health check $healthCheck to instance appId:${applicationKey.appId} version:${applicationKey.version}")
+    }
+
+    def removeHealthCheck(applicationKey: ApplicationKey, healthCheck: HealthCheck): Unit = {
+      logger.debug(s"Remove health check $healthCheck from instance appId:${applicationKey.appId} version:${applicationKey.version}")
+
+      val healthChecksAfterRemoval = healthChecks.getOrElse(applicationKey, Set.empty) - healthCheck
+      if (healthChecksAfterRemoval.isEmpty)
+        healthChecks.remove(applicationKey)
+      else
+        healthChecks.update(applicationKey, healthChecksAfterRemoval)
+
+      healthCheckStates = healthCheckStates.map(kv => {
+        val newHealthChecks = kv._2.filter(x => x._1 != healthCheck)
+        (kv._1, newHealthChecks)
+      }).filter(kv => kv._2.nonEmpty)
```

Comment:
```
healthCheckStates.keysIterator.foreach { k =>
  val newValue = healthCheckStates(k).filter { case (_, state) => state != healthCheck }
  if (newValue.isEmpty)
    healthCheckStates.remove(k)
  else
    healthCheckStates(k) = newValue
}
```

Reply:
You could also potentially implement a generic garbage collection method, which is called after `purgeHealthCheckStatuses`?

Reply:
I've implemented a generic purge method that is used to purge both HC definitions and HC results.